src/function/execute.rs

```diff
@@ -38,53 +36,114 @@ where
+
+        loop {
+            let active_query = zalsa_local.push_query(database_key_index);
+
+            // If we already executed this query once, then use the tracked-struct ids from the
+            // previous execution as the starting point for the new one.
+            if let Some(old_memo) = &opt_old_memo {
+                active_query.seed_tracked_struct_ids(&old_memo.revisions.tracked_struct_ids);
+            }
+
+            // Query was not previously executed, or value is potentially
+            // stale, or value is absent. Let's execute!
+            let mut new_value = C::execute(db, C::id_to_input(db, id));
+            let mut revisions = active_query.pop();
+
+            // If the new value is equal to the old one, then it didn't
+            // really change, even if some of its inputs have. So we can
+            // "backdate" its `changed_at` revision to be the same as the
+            // old value.
+            if let Some(old_memo) = &opt_old_memo {
+                self.backdate_if_appropriate(old_memo, &mut revisions, &new_value);
+                self.diff_outputs(db, database_key_index, old_memo, &mut revisions);
+            }
+
+            // Did the new result we got depend on our own provisional value, in a cycle?
+            if revisions.cycle_heads.contains(&database_key_index) {
+                let opt_owned_last_provisional;
+                let last_provisional_value = if let Some(last_provisional) = opt_last_provisional {
+                    // We have a last provisional value from our previous time around the loop.
+                    &last_provisional
+                        .value
+                        .as_ref()
+                        .expect("provisional value evicted by LRU?")
+                } else {
+                    // This is our first time around the loop; a provisional value must have been
+                    // inserted into the memo table when the cycle was hit, so let's pull our
+                    // initial provisional value from there.
+                    opt_owned_last_provisional = self.get_memo_from_table_for(zalsa, id);
+                    &opt_owned_last_provisional
+                        .as_deref()
+                        .expect(
+                            "{database_key_index:#?} is a cycle head, \
+                            but no provisional memo found",
+                        )
+                        .value
+                        .as_ref()
+                        .expect("provisional value evicted by LRU?")
+                };
                 tracing::debug!(
-                    "{database_key_index:?}: caught cycle {cycle:?}, have strategy {:?}",
-                    C::CYCLE_STRATEGY
+                    "{database_key_index:?}: execute: \
+                            I am a cycle head, comparing last provisional value \
+                            {last_provisional_value:#?} with new value {new_value:#?}"
```

Comment:
is value the user provided value? If so, you might not want to log that with tracing because it could be huge (e.g. an entire AST).

Reply:
Yeah makes sense. On the flip side, it can be super useful for debugging cycle iteration. I'll remove it for now and we'll see.